src/main/java/org/truffleruby/core/encoding/EncodingNodes.java

```diff
@@ -734,10 +734,22 @@ private RubyEncoding createDummy(String name) {
 
     @Primitive(name = "encoding_get_encoding_by_index", lowerFixnum = 0)
     public abstract static class GetEncodingObjectByIndexNode extends PrimitiveArrayArgumentsNode {
-        @Specialization
-        protected RubyEncoding getEncoding(int index) {
+
+        @Specialization(guards = "index == cachedIndex", limit = "getCacheLimit()")
```

Comment:
This needs a `isSingleContext()` guard, as with multi-contexts, encoding indices could correspond to different encodings.

Reply:
Thanks. I've added that and rebased the PR against the latest _master_.